Cassandra のバックアップ&リストアを動作確認できるようにする

xirc opened the pull request:
## 目標
`modules/service/redhat/core/resources/cassandra/backup/***.sh`
に含まれるCassandra バックアップツールをサンプルプロジェクトで動作確認ができるようにすること。

## 変更点
- `modules/service/redhat/core/resources/cassandra/backup/APP_cassandra_backup_kick.sh` を調整する
  - タイポ修正
  - マジックナンバーの変数化
  - tar のオプション順序を調整する
- サンプルプロジェクトに Cassandra バックアップに必要なユーザやサーバを追加する
- ~~README に免責事項の記載を追加する~~
  LICENSE にて言及しているため追加しないほうがよい
- Cassandra バックアップ方法をドキュメントにする
- Cassandra リストア方法をドキュメントにする

## 特に重点的に見てもらいところ
- バックアップ手順とその前提条件に誤りがないか
- リストア手順とその前提条件に誤りが無いか

Commit:
Use "${SSH_USER}" variable instead of a magic value

Commit:
Reconcile the order of tar options

"--null" should be specified at first.

> tar: The following options were used after any non-optional arguments in archive create or update mode.  These options are positional and affect only arguments that follow them.  Please, rearrange them properly.
> tar: --null has no effect
> tar: Exiting with failure status due to previous errors

Commit:
Use DR_HOST instead of PROD_HOST

It might be typo.

Commit:
Make it easy to test cassandra backup tools

- Add NFS server that is used by the backup tools
- Add terraform scripts to launch setup for the backup tools

Commit:
Write a guide to use cassandra backup tools

Commit:
Write a guide to restore from backuped cassandra data

Commit:
Add disclaimer statements

xirc commented on examples/aws_ec2/facility-cassandra-backup.tf:
```diff
@@ -0,0 +1,122 @@
+
+    : Replace the entry /apl/cassandra_backup of /etc/fstab
+    sudo sed -e '/\/apl\/cassandra_backup/d' -i.bak /etc/fstab
+    echo '${module.nfs_instance.private_ip}:${module.nfs_instance.nfs_export_path} /apl/cassandra_backup nfs defaults,noauto,user 0 0' | sudo tee /etc/fstab
```
[NFS クライアントの設定](https://access.redhat.com/documentation/ja-jp/red_hat_enterprise_linux/7/html/storage_administration_guide/nfs-clientconfig) に従い `/etc/fstab` を設定しています。

```
$ man fstab


       The fourth field (fs_mntops).
              This field describes the mount options associated with the filesystem.

              It is formatted as a comma-separated list of options.  It contains at least the type of mount (ro or rw), plus any additional options appropriate to the filesystem type (including performance-tuning options).  For details, see mount(8) or swapon(8).

              Basic filesystem-independent options are:

              defaults
                     use default options: rw, suid, dev, exec, auto, nouser, and async.

              noauto do not mount when "mount -a" is given (e.g., at boot time)

              user   allow a user to mount

              owner  allow device owner to mount

              comment
                     or x-<name> for use by fstab-maintaining programs

              nofail do not report errors for this device if it does not exist.

       The fifth field (fs_freq).
              This field is used by dump(8) to determine which filesystems need to be dumped.  Defaults to zero (don't dump) if not present.

       The sixth field (fs_passno).
              This  field  is used by fsck(8) to determine the order in which filesystem checks are done at boot time.  The root filesystem should be specified with a fs_passno of 1.  Other filesystems should have a fs_passno of 2.  Filesystems within a drive will
              be checked sequentially, but filesystems on different drives will be checked at the same time to utilize parallelism available in the hardware.  Defaults to zero (don't fsck) if not present.

```

バックアップスクリプトでは、`reactivejob` ユーザが、毎回マウント/アンマウントを実施するため、
マウントオプション(4番目のフィールド)には、`noauto`,`user` を追加する必要があります。

5番目、6番目のフィールドはデフォルト値です。
サンプルのため、デフォルト値で十分だと思っています。

negokaz replied:
良さそうです 👍

xirc commented on README.md:
```diff
@@ -68,7 +68,12 @@ lerna-terraform is released under the terms of the [Apache License Version 2.0](
 
+## 免責事項
+[Lerna Team] は、本リポジトリに含まれる内容について正確性には最大限の注意を払っておりますが、  
+本リポジトリに記載の情報を用いて行う一切の行為について、何らの責任を負うものではありません。
```
バックアップやリストアは特に重要な作業になるため、免責事項を追加しました。
記載している文で問題がある場合には教えて頂けると嬉しいです。

negokaz replied:
@xirc 

Apache License 2.0 にも免責の条項があるみたいなので、それと矛盾しないように過不足なく表現できているとより良いかなと思いました。

「Lerna Team **及び lerna-terraform のコントリビューター**は、本リポジトリに含まれる内容について正確性には最大限の注意を払っておりますが、本リポジトリ**の成果物を**用いて行う一切の行為について、何らの責任を負うものではありません。」

- コミッターとコントリビューター両方に責任がないことを明記
- 「記載の情報」はドキュメントに限定されている感があるので「成果物」と表記
    - ソースコードや（今はないですが）バイナリデータも「記載の情報」とみなして良いのかが少し自信持てませんでした

> 8. 責任の制限
> 
> いかなる条件および法理論においても、不法行為（過失を含む）、契約、またはその他いかなる場合でも、適用される法律または書面での同意によって命じられない限り、コントリビューターは本ライセンスまたは成果物の使い方に関連して生じる直接損害、間接損害、偶発的な損害、特別損害、懲罰的損害、または結果損害を含め、営業権の損失、業務の停止、コンピューター障害または誤作動、その他の商業上の損害や損失など、いかなる損害に対しても、たとえそうした損害の可能性をたとえ知らされていたとしても、あなたに責任を負わないものとします。
>
> **[Apache-2.0](https://licenses.opensource.jp/Apache-2.0/Apache-2.0.html#:~:text=8,%E3%80%82)** （[原文](https://opensource.org/licenses/Apache-2.0#:~:text=8.,damages.)）

xirc replied:
@negokaz 
ありがとうございます！
なるほど。ライセンスに明記されていることを失念していました。

> Apache License 2.0 にも免責の条項があるみたいなので、それと矛盾しないように過不足なく表現できているとより良いかなと思いました。

同意ですが、さらに踏み込んで、

Apache License 2.0 に免責事項が記載されているので、README などには**免責事項を書かない方**がより良さそう
に思いました。矛盾があると困るためです。

免責事項を追加したコミットを revert しようと思いますがどうでしょうか？ 
> Add disclaimer statements 52d2f3e

negokaz replied:
@xirc 

> Apache License 2.0 に免責事項が記載されているので、README などには免責事項を書かない方がより良さそうに思いました。矛盾があると困るためです。
> 
> 免責事項を追加したコミットを revert しようと思いますがどうでしょうか？

確かに法的に矛盾がないか検証するのも大変ですし、消してしまっても良さそうですね。
トップの README からライセンスの条文にアクセスできるようになっていますし、問題ないでしょう。

xirc replied:
@negokaz 
免責事項の記載を削除しました 😄 
> Revert "Add disclaimer statements" … 1038571

xirc commented on examples/aws_ec2/facility-cassandra-backup.tf:
```diff
@@ -0,0 +1,122 @@
+  provisioner "file" {
+    content     = file(var.ssh_public_key_file_path)
+    destination = "~/cassandra_backup_public_key"
+  }
```
ssh key pair は バックアップ用ユーザ `reactivejob` 固有で作成したほうが良さそうですが、
サンプルとしての簡潔さを保つため共通のカギを使っています。

negokaz replied:
@xirc 確かにサンプルで個別に鍵作るとユーザーが試す上で手間が増えそうですね。

ただ、このサンプルを流用して本番向けのリソースを作成することも考えられるので、コメントで「`reactivejob` 固有の鍵を作ることを推奨します」と注意を促したほうがより親切かなと思いました。

xirc replied:
> https://github.com/lerna-stack/lerna-terraform/pull/6#discussion_r596569023

手間は増えますが、`reactivejob` の秘密鍵&公開鍵をユーザに作成してもらうようにします。

xirc replied:
次のディスカッションに関連しています。

> https://github.com/lerna-stack/lerna-terraform/pull/6#discussion_r596658455
> reactivejob と centos が使用する鍵を別にしました。
> centos で使う鍵のデフォルト値はこれまで通り ~/.ssh/id_rsa です。
> reactivejob で使う鍵は ./resources/id_rsa_cassandra_backup_user にしました。
> reactivejob 向けの鍵を作る手順は README に記載しました。

手間は増えますが、`reactivejob` のキーペアをユーザに作成してもらうようにしました。
この秘密鍵は Cassandra サーバ間で共有されることも README に記載しました。

xirc commented on examples/aws_ec2/modules/nfs-instance/main.tf:
```diff
@@ -0,0 +1,72 @@
+    sudo chmod 777 "${var.nfs_export_path}"
+
+    sudo systemctl restart nfs-server
+    sudo systemctl enable nfs-server
```
NFS4 では portmap、rpcbind、nfslock は不要でした。
`nfs-utils` をインストールし、`nfs-server` を起動するだけです 😃 

## 参考
- [NFSサーバーの構築ついて纏めてみる](https://qiita.com/leomaro7/items/bb1b46693614b05df7bd)
- [NFS サーバーの起動と停止](https://access.redhat.com/documentation/ja-jp/red_hat_enterprise_linux/7/html/storage_administration_guide/s1-nfs-start)
- [NFS サーバーの設定](https://access.redhat.com/documentation/ja-jp/red_hat_enterprise_linux/7/html/storage_administration_guide/nfs-serverconfig)

negokaz replied:
@xirc 

RHEL のドキュメントには  NFS4 専用とするケースでは次のような手順が必要と記載されていましたが、これは対象とはならない感じでしょうか？
RHEL のドキュメントではサービス名が `nfs` なので、そもそもサービスが全然違う？

> /etc/sysconfig/nfs 設定ファイルに以下の行を追加して、NFSv2、NFSv3、UDP を無効化します。


> 8.6.7. NFSv4 専用サーバーの設定
>
> **[8.6. NFS サーバーの設定 Red Hat Enterprise Linux 7 | Red Hat Customer Portal](https://access.redhat.com/documentation/ja-jp/red_hat_enterprise_linux/7/html/storage_administration_guide/nfs-serverconfig#nfs4-only)**

xirc replied:
@negokaz

NFS には詳しくありませんが、  
コメント頂いたものは NFS4 **のみ** をサポートしてほかのプロトコルを使用させないための手順に読めるので、
恐らく不要だと思います。


> [手順8.2 NFSv4 専用サーバーの設定](https://access.redhat.com/documentation/ja-jp/red_hat_enterprise_linux/7/html/storage_administration_guide/nfs-serverconfig#nfs4-only)
> NFS サーバーを NFS バージョン 4.0 以降のみをサポートするように設定する方法を説明します。

サンプルに含めるプライベートネットワークに配置するサーバなので細かい検証はせず、
NFSサーバとしてのファイルがホストができることしか確認していません。
設定項目等を検証したほうがよろしいでしょうか？

negokaz replied:
@xirc 確かに、サンプルなのでファイルの読み書きができることが確認できていれば問題なさそうです。
追加の検証はなしでいきましょう。

xirc commented on examples/aws_ec2/modules/nfs-instance/main.tf:
```diff
@@ -0,0 +1,72 @@
+
+  # Only For Development
+  # It would be better to use more strict access control.
+  ${var.nfs_export_path} (rw)
```
サンプルとしてのシンプルさを保つため、
このサーバにアクセスできるユーザは 読み書き できるようになっています。

negokaz replied:
良いと思います 👍

xirc commented on examples/aws_ec2/versions.tf:
```diff
@@ -10,3 +10,9 @@ terraform {
+  region     = "ap-northeast-1"
+  access_key = var.aws_access_key
+  secret_key = var.aws_secret_key
+}
```
`examples/aws_ec2/modules/nfs-instance` で使用されます。  
個別のモジュールではなく、サンプル全体で指定するほうが良さそうでした。

> [Providers Within Modules](https://www.terraform.io/docs/language/modules/develop/providers.html)
> A module intended to be called by one or more other modules must not contain any provider blocks, with the exception of the special "proxy provider blocks" discussed under Passing Providers Explicitly below.

Commit:
Provide default AMI, ssh_user, and tags

Commit:
Write README of modules/nfs-instance

negokaz commented on examples/aws_ec2/variables.tf:
```diff
@@ -44,6 +44,18 @@ variable "ssh_user" {
+variable "ssh_private_key_file_path" {
+  description = "SSH 秘密鍵 のファイルパス"
+  type        = string
+  default     = "~/.ssh/id_rsa"
```
@xirc 

この秘密鍵はサーバー上で他のメンバーがアクセスできる状態になる可能性があるので、その旨注意書きを `description` に書いておいたほうが良さそうです。

**例：**「この秘密鍵はサーバー上に配置されます。サーバーにアクセスできるメンバー全員がこの秘密鍵にアクセスj可能になるため、専用の秘密鍵を新規作成することをお勧めします。」

また、`~/.ssh/id_rsa` は `ssh_keygen` で生成される鍵のデフォルト名であり他の用途で利用している可能性が高そうなので、`~/.ssh/id_rsa_lerna_terraform_sample`  などをデフォルトにしておくのが良さそうに思いました。

xirc replied:
@negokaz

そうですね。

デフォルト値を元のコードから変えないように設定を移動しましたが、デフォルト値は変えた方がよさそうです。
また、ユーザの ホーム `~/` をデフォルトで参照するのは良くないように思います。
デフォルトのカギの場所は `${path.module}/resources/id_rsa` に移動しますがどうでしょうか？
公開鍵の場所も合わせて `${path.module}/resource/id_rsa.pub` に移動します。
ついでに、ユーザ `reactivejob` 用のカギと、ユーザ `centos` のカギを別にします。

注意書きも書いておく方が良さそうです。
また、ユーザに `ssh-keygen` でカギを作ってもらうように README に手順を記載します。

negokaz replied:
@xirc 

安全側に倒す感じですね。良いと思います。

`reactivejob` の鍵は `resource` 配下で問題ないと思います。`centos` の鍵は  SSH クライアントが参照しにいくので、`~/.ssh` 配下のほうがいろいろ便利かなと思いました。Terraform プロジェクトのディレクトリをリネームしたりするとクライアントが鍵を参照できなくなってしまいますし、確か Linux や Mac だと鍵があるディレクトリのパーミッションが緩いと `ssh` コマンドで繋ぐときにエラーになってしまったと思うので。

xirc replied:
@negokaz 

> Manage keypairs, individually … 3130ce4

`reactivejob` と `centos` が使用する鍵を別にしました。
`centos` で使う鍵のデフォルト値はこれまで通り `~/.ssh/id_rsa` です。  
`reactivejob` で使う鍵は `./resources/id_rsa_cassandra_backup_user` にしました。
`reactivejob` 向けの鍵を作る手順は README に記載しました。


> Add triggers to facility-cassandra-backup … 03787f4

併せて、動作確認等で鍵を更新しやすいように trigger を追加しました。

xirc replied:
> Add a caution about the backup user's private key 22f1ea2

鍵に関する注意事項を README にも記載しました。

negokaz replied:
@xirc 

`generate-variables-example.sh` がヒアドキュメント記法をサポートしていないので check が落ちるようになってしまいました…

```
❯ bin/generate-variables-example.sh check-tfvars examples/aws_ec2/variables.tf
ファイル examples/aws_ec2/terraform.tfvars.example と /dev/fd/62 は異なります

examples/aws_ec2/variables.tf が変更されているため、terraform.tfvars.example の更新が必要です。
'generate-variables-example.sh generate-tfvars "examples/aws_ec2/variables.tf"' を実行してください。
```

CI でのチェックが難しくなるので `cassandra_backup_user_ssh_private_key_filepath` の `description` を改行せずに 1 行にしませんか？1 行が長くなってしまいますが、ぎりぎり許容範囲かなと思います。

ツールのほうを修正できれば理想的ですが、すぐには対応難しそうです。

xirc replied:
@negokaz 
> Use an one-line string instead of a here document … 304a522

変更しました。

negokaz commented on docs/ops/Cassandraリストア.md:
````diff
@@ -0,0 +1,185 @@
+
+Cassandra サービスを停止するため、次のコマンドを実行してください。
+```shell
+$ sudo systemctl stop cassandra
````
@xirc 

この手順はおそらく全てのクライアントからの接続を閉じるためのものだと思いますが、`nodetool` などを使って接続されているクライアント数を確認する方法があれば、その確認をしたほうが確実かなと思いました。
再接続してくるクライアントが居たときに、それに気付かず接続があるままリストアを実行してしまうことになるので。

クライアント数の確認方法がないか少し調査していただくことは可能ですか？
前提条件に「サービスを全て停止していること」と記載されているので、10分ぐらい調べて見つからなければ無しでも問題ないかなと思っています。

xirc replied:
> クライアント数の確認方法がないか少し調査していただくことは可能ですか？

承知しました。少し調べてみましょう。

xirc replied:
> [How to know Native Clients connected to Cassandra](https://stackoverflow.com/questions/54212510/how-to-know-native-clients-connected-to-cassandra)
> `sudo lsof -i -n -P | grep 9042 | grep ESTABLISHED`

`lsof` や `netstat` を使えば確認できそうですね。

> [nodetool clientstats --all](https://cassandra.apache.org/doc/latest/tools/nodetool/clientstats.html)

Cassandra 4.x になると `nodetool clientstats` が使えそうです。

```shell
$ netstat --tcp -a | grep 9042
tcp        0      0 ip-10-0-1-61.ap-no:9042 0.0.0.0:*               LISTEN
tcp        0     25 ip-10-0-1-61.ap-no:9042 10.0.1.51:44638         ESTABLISHED
tcp        0      0 ip-10-0-1-61.ap-no:9042 10.0.1.51:44772         ESTABLISHED
tcp        0      0 ip-10-0-1-61.ap-no:9042 10.0.1.51:44768         ESTABLISHED
tcp        0      0 ip-10-0-1-61.ap-no:9042 10.0.1.51:44752         ESTABLISHED
tcp        0      0 ip-10-0-1-61.ap-no:9042 10.0.1.51:44742         ESTABLISHED
tcp        0      0 ip-10-0-1-61.ap-no:9042 10.0.1.51:44758         ESTABLISHED
tcp        0      0 ip-10-0-1-61.ap-no:9042 10.0.1.51:44744         ESTABLISHED
tcp        0     25 ip-10-0-1-61.ap-no:9042 10.0.1.51:44640         ESTABLISHED
tcp        0      0 ip-10-0-1-61.ap-no:9042 10.0.1.51:44740         ESTABLISHED
tcp        0      0 ip-10-0-1-61.ap-no:9042 10.0.1.51:44632         ESTABLISHED
```

アプリサーバが `10.0.1.51` で、 Cassandra ノード `10.0.1.61` で実行しました。
すべてのノードにログインして確認すれば良さそうです。

xirc replied:
@negokaz 

> Add descriptions of how to check there are no Cassandra clients bc41ff7

`netstat` を使って確認する方法を記載しました。

negokaz replied:
良さそうです  👍

negokaz commented on docs/ops/Cassandraバックアップ.md:
```diff
@@ -0,0 +1,79 @@
+  バックアップスクリプトを起動する Cassandra サーバにインストールされている `/opt/management/bin/APP_cassandra_backup_kick.sh` の変数を手動で書き換える必要があります。
+- 1つのCassandraノードのみをバックアップ対象とします。  
+  Cassandra クラスタに参加しているすべてのノードにデータがレプリケーションされていることを想定しています。  
+  キースペースの レプリケーションファクタ が Cassandra クラスタのノード数よりも小さい場合はサポートしていません。
```
@xirc 

バックアップスクリプトは過去 3 世代 + 最新 1 世代の計 4 世代分のバックアップを残すように構成されていると思うので、そのことについても触れたほうが良いかもしれません。

> ```bash
> ###############################
> #generation
> ###############################
> function rotate_archives {
>   # コマンドグループ
>   {
>       if [[ ${TENANT_ID} == ${TENANT_ID_WITH_LEGACY_BACKUP_FILE_FORMAT} ]] ; then
>           /bin/find ${NODE_BACKUP_DIR} -maxdepth 1 -name "${BASE_FILE_NAME}_[0-9][0-9][0-9][0-9]*" -or -name "${BASE_FILE_NAME}_${TENANT_ID}_*"
>       else
>           /bin/find ${NODE_BACKUP_DIR} -maxdepth 1 -name "${BASE_FILE_NAME}_${TENANT_ID}_*"
>       fi
> 
>   } | xargs ls -t -dF |tail --lines=+$((GENERATION_DAIRY+1)) | xargs -I% rm -rf % ;RC=${?}
>   if [[ ${RC} -ne 0 ]] ; then
>       log ${RC} "generation ${EXECUTE_HOST} is abnormal end."
>       exit 1
>   else
>       log ${RC} "generation ${EXECUTE_HOST} is success."
>   fi
> }
> ```
> **[lerna-terraform/APP_cassandra_backup_kick.sh at yamakawa/cassandra-backup · lerna-stack/lerna-terraform](https://github.com/lerna-stack/lerna-terraform/blob/28e4727042545938d9a165accd00d457dac8f2f0/modules/service/redhat/core/resources/cassandra/backup/APP_cassandra_backup_kick.sh#L241-L260)**

xirc replied:
@negokaz 
> Add descriptions about a backup rotation c0c26c6

バックアップファイルのローテーション仕様を記載しました。

negokaz replied:
@xirc 

>   バックアップスクリプトには、何世代分のログを残すかを設定する変数がありますが、この変数を書き換えることは非推奨です。

これは、Terraform によって上書きされる可能性があるからですかね？

xirc replied:
> これは、Terraform によって上書きされる可能性があるからですかね？

はい、そうです。

xirc replied:
@negokaz 
利用者によるスクリプトの書き換えがあるのか、を見直してみました。
バックアップを取得するには、利用者が手動でスクリプトを書き換える必要がありました。

> https://github.com/lerna-stack/lerna-terraform/pull/6/files/28e4727042545938d9a165accd00d457dac8f2f0#diff-1c437ea2477b86000343cec75b1adcffe462c58d6c668488c383c425cc71b65bR20
>  `/opt/management/bin/APP_cassandra_backup_kick.sh` にバックアップ対象の Cassandra サーバの IP アドレスが変数(`PROD_HOSTS`, `DR_HOSTS`) にハードコードされています。  
  バックアップスクリプトを起動する Cassandra サーバにインストールされている `/opt/management/bin/APP_cassandra_backup_kick.sh` の変数を手動で書き換える必要があります。

ローテーション以外の箇所で利用者の書き換えが必要になっているめ、次の説明は無い方が良さそうに思えます。
次の1文を削除しようと思いますがどうでしょうか？

> バックアップスクリプトには、何世代分のログを残すかを設定する変数がありますが、この変数を書き換えることは非推奨です。

negokaz replied:
@xirc 

> ローテーション以外の箇所で利用者の書き換えが必要になっているめ、次の説明は無い方が良さそうに思えます。
> 次の1文を削除しようと思いますがどうでしょうか？

なるほど。良いと思います 👍

xirc replied:
@negokaz 
> Remove the unnecessary statement d9cab3e

削除しました 😄

Commit:
Revert "Add disclaimer statements"

This reverts commit 52d2f3ecee8eb7f3ac316d253900808a0c3cf58b.

Commit:
Manage keypairs, individually

It would be better to manage key pairs individually since a backup user's ssh private key is exposed to teammates.
Tammates can access the private key since it is distributed among Cassandra servers.

Commit:
Add triggers to facility-cassandra-backup

To update ssh keys easily

Commit:
Add descriptions about a backup rotation

Commit:
Add a caution about the backup user's private key

Commit:
Add descriptions of how to check there are no Cassandra clients

Commit:
Use an one-line string instead of a here document

CI bash scripts doesn't support a here document.

Commit:
Remove the unnecessary statement
